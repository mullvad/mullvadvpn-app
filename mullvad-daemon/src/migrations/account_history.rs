use super::{Error, Result};
use mullvad_types::account::AccountToken;
use once_cell::sync::Lazy;
use regex::Regex;
use serde::Deserialize;
use std::path::Path;
use talpid_types::ErrorExt;
use tokio::{
    fs::{self, File},
    io::{self, AsyncReadExt, AsyncSeekExt, AsyncWriteExt},
};

// ======================================================
// Section for vendoring types.

// ======================================================

const ACCOUNT_HISTORY_FILE: &str = "account-history.json";

static ACCOUNT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[0-9]+$").unwrap());

pub async fn migrate_location(old_dir: &Path, new_dir: &Path) {
    let old_path = old_dir.join(ACCOUNT_HISTORY_FILE);
    let new_path = new_dir.join(ACCOUNT_HISTORY_FILE);
    if !old_path.exists() || new_path.exists() || new_path == old_path {
        return;
    }

    if let Err(error) = fs::copy(&old_path, &new_path).await {
        log::error!(
            "{}",
            error.display_chain_with_msg("Failed to migrate account history file location")
        );
    } else {
        let _ = fs::remove_file(old_path).await;
    }
}

pub async fn migrate_formats(settings_dir: &Path, settings: &mut serde_json::Value) -> Result<()> {
    let path = settings_dir.join(ACCOUNT_HISTORY_FILE);
    if !path.is_file() {
        return Ok(());
    }

    let mut options = fs::OpenOptions::new();
    #[cfg(unix)]
    {
        options.mode(0o600);
    }
    let mut file = options
        .write(true)
        .read(true)
        .open(path)
        .await
        .map_err(Error::ReadHistory)?;

    let mut bytes = vec![];
    file.read_to_end(&mut bytes)
        .await
        .map_err(Error::ReadHistory)?;

    if is_format_v3(&bytes) {
        return Ok(());
    }
    write_format_v3(file, migrate_formats_inner(&bytes, settings)?).await
}

fn migrate_formats_inner(
    account_bytes: &[u8],
    settings: &mut serde_json::Value,
) -> Result<Option<AccountToken>> {
    if let Ok(result) = try_format_v2(account_bytes) {
        if let Some((token, wg_data)) = result {
            settings["wireguard"] = wg_data;
            Ok(Some(token))
        } else {
            Ok(None)
        }
    } else if let Ok(token) = try_format_v1(account_bytes) {
        Ok(token)
    } else {
        Err(Error::ParseHistoryError)
    }
}

fn is_format_v3(bytes: &[u8]) -> bool {
    match std::str::from_utf8(bytes) {
        Ok(token) => token.is_empty() || ACCOUNT_REGEX.is_match(token),
        Err(_) => false,
    }
}

async fn write_format_v3(mut file: File, token: Option<AccountToken>) -> Result<()> {
    file.set_len(0).await.map_err(Error::WriteHistory)?;
    file.seek(io::SeekFrom::Start(0))
        .await
        .map_err(Error::WriteHistory)?;
    if let Some(token) = token {
        file.write_all(token.as_bytes())
            .await
            .map_err(Error::WriteHistory)?;
    }
    file.sync_all().await.map_err(Error::WriteHistory)
}

fn try_format_v2(bytes: &[u8]) -> Result<Option<(AccountToken, serde_json::Value)>> {
    #[derive(Deserialize, Clone)]
    pub struct AccountEntry {
        pub account: AccountToken,
        pub wireguard: serde_json::Value,
    }
    Ok(serde_json::from_slice::<'_, Vec<AccountEntry>>(bytes)
        .map_err(|_error| Error::ParseHistoryError)?
        .into_iter()
        .next()
        .map(|entry| (entry.account, entry.wireguard)))
}

fn try_format_v1(bytes: &[u8]) -> Result<Option<AccountToken>> {
    #[derive(Deserialize)]
    struct OldFormat {
        accounts: Vec<AccountToken>,
    }
    Ok(serde_json::from_slice::<'_, OldFormat>(bytes)
        .map_err(|_error| Error::ParseHistoryError)?
        .accounts
        .into_iter()
        .next())
}

#[cfg(test)]
mod test {

    pub const ACCOUNT_HISTORY_V1: &str = r#"
{
  "accounts": ["1234", "4567"]
}
"#;
    pub const ACCOUNT_HISTORY_V1_EMPTY: &str = r#"
{
  "accounts": []
}
"#;
    pub const ACCOUNT_HISTORY_V2: &str = r#"
[
  {
  "account": "1234",
    "wireguard": {
      "private_key": "mAdSb4AfQOsAD5O/5+zG1oIhk3cUl0jUsyOeaOMFu3o=",
      "addresses": {
        "ipv4_address": "109.111.108.101/32",
        "ipv6_address": "ffff::ffff/128"
      },
      "created": "1970-01-01T00:00:00Z"
    }
  },
  {
    "account": "4567",
    "wireguard": {
      "private_key": "mAdSb4AfQOsAD5O/5+zG1oIhk3cUl0jUsyOeaOMFu3o=",
      "addresses": {
        "ipv4_address": "109.111.108.101/32",
        "ipv6_address": "ffff::ffff/128"
      },
      "created": "1970-01-01T00:00:00Z"
    }
  }
]"#;
    pub const ACCOUNT_HISTORY_V2_EMPTY: &str = r#"[]"#;
    pub const ACCOUNT_HISTORY_V3: &str = r#"123456"#;

    pub const OLD_SETTINGS: &str = r#"
{
  "account_token": "1234",
  "relay_settings": {
    "normal": {
      "location": {
        "only": {
          "country": "se"
        }
      },
      "tunnel_protocol": "any",
      "wireguard_constraints": {
        "port": {
          "only": {
            "protocol": "tcp",
            "port": {
              "only": 80
            }
          }
        }
      },
      "openvpn_constraints": {
        "port": {
          "only": {
            "protocol": "udp",
            "port": {
              "only": 1195
            }
          }
        }
      }
    }
  },
  "bridge_settings": {
    "normal": {
      "location": "any"
    }
  },
  "bridge_state": "auto",
  "allow_lan": true,
  "block_when_disconnected": false,
  "auto_connect": false,
  "tunnel_options": {
    "openvpn": {
      "mssfix": null
    },
    "wireguard": {
      "mtu": null,
      "rotation_interval": {
          "secs": 86400,
          "nanos": 0
      }
    },
    "generic": {
      "enable_ipv6": false
    },
    "dns_options": {
      "state": "default",
      "default_options": {
        "block_ads": false,
        "block_trackers": false
      },
      "custom_options": {
        "addresses": [
          "1.1.1.1",
          "1.2.3.4"
        ]
      }
    }
  },
  "settings_version": 5
}
"#;

    pub const NEW_SETTINGS: &str = r#"
{
  "account_token": "1234",
  "wireguard": {
    "private_key": "mAdSb4AfQOsAD5O/5+zG1oIhk3cUl0jUsyOeaOMFu3o=",
    "addresses": {
      "ipv4_address": "109.111.108.101/32",
      "ipv6_address": "ffff::ffff/128"
    },
    "created": "1970-01-01T00:00:00Z"
  },
  "relay_settings": {
    "normal": {
      "location": {
        "only": {
          "country": "se"
        }
      },
      "tunnel_protocol": "any",
      "wireguard_constraints": {
        "port": {
          "only": {
            "protocol": "tcp",
            "port": {
              "only": 80
            }
          }
        }
      },
      "openvpn_constraints": {
        "port": {
          "only": {
            "protocol": "udp",
            "port": {
              "only": 1195
            }
          }
        }
      }
    }
  },
  "bridge_settings": {
    "normal": {
      "location": "any"
    }
  },
  "bridge_state": "auto",
  "allow_lan": true,
  "block_when_disconnected": false,
  "auto_connect": false,
  "tunnel_options": {
    "openvpn": {
      "mssfix": null
    },
    "wireguard": {
      "mtu": null,
      "rotation_interval": {
          "secs": 86400,
          "nanos": 0
      }
    },
    "generic": {
      "enable_ipv6": false
    },
    "dns_options": {
      "state": "default",
      "default_options": {
        "block_ads": false,
        "block_trackers": false
      },
      "custom_options": {
        "addresses": [
          "1.1.1.1",
          "1.2.3.4"
        ]
      }
    }
  },
  "settings_version": 5
}
"#;

    // Test whether the current format is parsed correctly
    #[test]
    fn test_v3() {
        assert!(!super::is_format_v3(ACCOUNT_HISTORY_V1.as_bytes()));
        assert!(!super::is_format_v3(ACCOUNT_HISTORY_V2.as_bytes()));
        assert!(super::is_format_v3(ACCOUNT_HISTORY_V3.as_bytes()));
    }

    #[test]
    fn test_v2() {
        assert!(super::try_format_v2(ACCOUNT_HISTORY_V1.as_bytes()).is_err());

        let mut old_settings = serde_json::from_str(OLD_SETTINGS).unwrap();
        let new_settings: serde_json::Value = serde_json::from_str(NEW_SETTINGS).unwrap();

        // Test whether the wireguard data is moved to the settings correctly
        let token =
            super::migrate_formats_inner(ACCOUNT_HISTORY_V2.as_bytes(), &mut old_settings).unwrap();

        assert_eq!(&old_settings, &new_settings);
        assert_eq!(token, Some("1234".to_string()));

        // Test whether empty histories are handled correctly
        let mut old_settings = serde_json::from_str(OLD_SETTINGS).unwrap();
        let token =
            super::migrate_formats_inner(ACCOUNT_HISTORY_V2_EMPTY.as_bytes(), &mut old_settings)
                .unwrap();
        assert_eq!(&old_settings, &old_settings);
        assert_eq!(token, None);
    }

    #[test]
    fn test_v1() {
        let token = super::try_format_v1(ACCOUNT_HISTORY_V1.as_bytes()).unwrap();
        assert_eq!(token, Some("1234".to_string()));
        let token = super::try_format_v1(ACCOUNT_HISTORY_V1_EMPTY.as_bytes()).unwrap();
        assert_eq!(token, None);
    }
}
