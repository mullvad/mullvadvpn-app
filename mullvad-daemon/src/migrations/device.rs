//! Generates a `device.json` from a WireGuard key and account number by matching them against
//! devices returned by the API and sending the `DeviceMigrationEvent` event to the daemon.
//! The account number and private key may be lost if it fails, but this should not be not
//! critical since the account history also contains the token.
//!
//! This module is allowed to import a number of types, unlike other migration modules, as it
//! does not modify any files directly and may safely fail.

use super::{MigrationComplete, v5::MigrationData};
use crate::{
    DaemonEventSender, InternalDaemonEvent,
    device::{self, DeviceService, PrivateAccountAndDevice, PrivateDevice},
};
use mullvad_types::{account::AccountNumber, wireguard::WireguardData};
use std::time::Duration;
use talpid_core::mpsc::Sender;
use talpid_types::ErrorExt;
use tokio::time::timeout;

const TIMEOUT: Duration = Duration::from_secs(30);

pub(crate) fn generate_device(
    migration_data: MigrationData,
    mut migration_complete: MigrationComplete,
    rest_handle: mullvad_api::rest::MullvadRestHandle,
    daemon_tx: DaemonEventSender,
) {
    tokio::spawn(async move {
        let wg_data: Option<WireguardData> = migration_data.wg_data.and_then(|data| {
            serde_json::from_value(data)
                .map(Some)
                .unwrap_or_else(|error| {
                    log::error!(
                        "{}",
                        error.display_chain_with_msg("Failed to parse WireGuard data")
                    );
                    None
                })
        });

        let api_handle = rest_handle.availability.clone();
        let service = DeviceService::new(rest_handle, api_handle);
        let result = match (migration_data.token, wg_data) {
            (account_number, Some(wg_data)) => {
                log::info!("Creating a new device cache from previous settings");
                cache_from_wireguard_key(service, account_number, wg_data).await
            }
            (account_number, None) => {
                log::info!("Generating a new device for the account");
                cache_from_account(service, account_number).await
            }
        };
        let _ = daemon_tx.send(InternalDaemonEvent::DeviceMigrationEvent(result));
        migration_complete.set_complete();
    });
}

async fn cache_from_wireguard_key(
    service: DeviceService,
    account_number: AccountNumber,
    wg_data: WireguardData,
) -> Result<PrivateAccountAndDevice, device::Error> {
    let devices = timeout(
        TIMEOUT,
        service.list_devices_with_backoff(account_number.clone()),
    )
    .await
    .map_err(|_error| {
        log::error!("Failed to enumerate devices for account: timed out");
        device::Error::Cancelled
    })?
    .inspect_err(|error| {
        log::error!(
            "{}",
            error.display_chain_with_msg("Failed to enumerate devices for account")
        );
    })?;

    for device in devices.into_iter() {
        if device.pubkey == wg_data.private_key.public_key() {
            return Ok(PrivateAccountAndDevice {
                account_number,
                device: PrivateDevice::try_from_device(device, wg_data)?,
            });
        }
    }
    log::info!("The existing WireGuard key is not valid");
    Err(device::Error::InvalidDevice)
}

async fn cache_from_account(
    service: DeviceService,
    account_number: AccountNumber,
) -> Result<PrivateAccountAndDevice, device::Error> {
    timeout(
        TIMEOUT,
        service.generate_for_account_with_backoff(account_number),
    )
    .await
    .map_err(|_error| {
        log::error!("Failed to generate new device for account: timed out");
        device::Error::Cancelled
    })?
    .inspect_err(|error| {
        log::error!(
            "{}",
            error.display_chain_with_msg("Failed to generate new device for account")
        );
    })
}
