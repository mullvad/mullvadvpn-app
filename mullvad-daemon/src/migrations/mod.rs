//! Code for migrating between different versions of the settings.
//! Migration only supports migrating forward, to newer formats.
//!
//! A settings migration module is responsible for converting
//! from its own version to the next version. So `v3::migrate`
//! migrates from settings version `V3` to `V4` etc.
//!
//! Migration modules may NOT import and use structs that may
//! change. Because then a later change to the current code can break
//! old migrations. The only items a settings migration module may import
//! are anything from `std`, `jnix`, `serde` and the following:
//!
//! ```ignore
//! use super::{Error, Result};
//! use mullvad_types::relay_constraints::Constraint;
//! use mullvad_types::settings::SettingsVersion;
//! ```
//!
//! Any other type must be vendored into the migration module so the format
//! it has is locked over time.
//!
//! There should never be multiple migrations between two official releases. At most one.
//! Between releases, dev builds can break the settings without having a proper migration path.
//!
//! # Creating a migration
//!
//! 1. Copy `vX.rs.template` to `vX.rs` where `X` is the latest settings version right now.
//! 1. Add the new version (`Y = X+1`) to `SettingsVersion` and bump `CURRENT_SETTINGS_VERSION`
//!    to `Y`.
//! 1. Write a comment in the new module about how the format changed, what it needs to migrate.
//! 1. Implement the migration and add adequate tests.
//! 1. Add to the changelog: "Settings format updated to `vY`"

use std::{
    path::Path,
    sync::{
        atomic::{AtomicBool, Ordering},
        Arc,
    },
};
use tokio::{
    fs,
    io::{self, AsyncWriteExt},
};

mod account_history;
mod device;
mod v1;
mod v2;
mod v3;
mod v4;
mod v5;
mod v6;

const SETTINGS_FILE: &str = "settings.json";

#[derive(err_derive::Error, Debug)]
#[error(no_from)]
pub enum Error {
    #[error(display = "Failed to read the settings")]
    Read(#[error(source)] io::Error),

    #[error(display = "Failed to deserialize settings")]
    Deserialize(#[error(source)] serde_json::Error),

    #[error(display = "Unexpected settings format")]
    InvalidSettingsContent,

    #[error(display = "Unable to serialize settings to JSON")]
    Serialize(#[error(source)] serde_json::Error),

    #[error(display = "Unable to open settings for writing")]
    Open(#[error(source)] io::Error),

    #[error(display = "Unable to write new settings")]
    Write(#[error(source)] io::Error),

    #[error(display = "Unable to sync settings to disk")]
    SyncSettings(#[error(source)] io::Error),

    #[error(display = "Failed to read the account history")]
    ReadHistory(#[error(source)] io::Error),

    #[error(display = "Failed to write new account history")]
    WriteHistory(#[error(source)] io::Error),

    #[error(display = "Failed to parse account history")]
    ParseHistoryError,

    #[cfg(windows)]
    #[error(display = "Failed to restore Windows update backup")]
    WinMigrationError(#[error(source)] windows::Error),
}

pub type Result<T> = std::result::Result<T, Error>;

/// Returns whether there is any background work remaining.
#[derive(Clone)]
pub struct MigrationComplete(Arc<AtomicBool>);

impl MigrationComplete {
    pub fn new(state: bool) -> Self {
        Self(Arc::new(AtomicBool::new(state)))
    }

    pub fn is_complete(&self) -> bool {
        self.0.load(Ordering::Relaxed)
    }

    fn set_complete(&mut self) {
        self.0.store(true, Ordering::Relaxed);
    }
}

/// Contains discarded data that may be useful for later work.
pub type MigrationData = v5::MigrationData;

pub async fn migrate_all(cache_dir: &Path, settings_dir: &Path) -> Result<Option<MigrationData>> {
    #[cfg(windows)]
    windows::migrate_after_windows_update(settings_dir)
        .await
        .map_err(Error::WinMigrationError)?;

    let path = settings_dir.join(SETTINGS_FILE);

    if !path.is_file() {
        return Ok(None);
    }

    let settings_bytes = fs::read(&path).await.map_err(Error::Read)?;

    let mut settings: serde_json::Value =
        serde_json::from_reader(&settings_bytes[..]).map_err(Error::Deserialize)?;

    if !settings.is_object() {
        return Err(Error::InvalidSettingsContent);
    }

    let old_settings = settings.clone();

    v1::migrate(&mut settings)?;
    v2::migrate(&mut settings)?;
    v3::migrate(&mut settings)?;
    v4::migrate(&mut settings)?;

    account_history::migrate_location(cache_dir, settings_dir).await;
    account_history::migrate_formats(settings_dir, &mut settings).await?;

    let migration_data = v5::migrate(&mut settings)?;
    v6::migrate(&mut settings)?;

    if settings == old_settings {
        // Nothing changed
        return Ok(migration_data);
    }

    let buffer = serde_json::to_string_pretty(&settings).map_err(Error::Serialize)?;

    let mut file = fs::OpenOptions::new()
        .create(true)
        .write(true)
        .truncate(true)
        .open(&path)
        .await
        .map_err(Error::Open)?;
    file.write_all(&buffer.into_bytes())
        .await
        .map_err(Error::Write)?;
    file.sync_data().await.map_err(Error::SyncSettings)?;

    log::debug!("Migrated settings. Wrote settings to {}", path.display());

    Ok(migration_data)
}

pub(crate) fn migrate_device(
    migration_data: MigrationData,
    rest_handle: mullvad_api::rest::MullvadRestHandle,
    daemon_tx: crate::DaemonEventSender,
) -> MigrationComplete {
    let migration_complete = MigrationComplete::new(false);
    device::generate_device(
        migration_data,
        migration_complete.clone(),
        rest_handle,
        daemon_tx,
    );
    migration_complete
}

#[cfg(windows)]
mod windows {
    use std::{ffi::OsStr, io, os::windows::ffi::OsStrExt, path::Path, ptr};
    use talpid_types::ErrorExt;
    use tokio::fs;
    use windows_sys::Win32::{
        Foundation::{ERROR_SUCCESS, HANDLE, PSID},
        Security::{
            Authorization::{GetNamedSecurityInfoW, SE_FILE_OBJECT, SE_OBJECT_TYPE},
            IsWellKnownSid, WinBuiltinAdministratorsSid, WinLocalSystemSid,
            OWNER_SECURITY_INFORMATION, SECURITY_DESCRIPTOR, SID, WELL_KNOWN_SID_TYPE,
        },
        System::Memory::LocalFree,
    };

    #[allow(non_camel_case_types)]
    type SECURITY_INFORMATION = u32;

    const MIGRATION_DIRNAME: &str = "windows.old";
    const MIGRATE_FILES: [(&str, bool); 3] = [
        ("settings.json", true),
        ("device.json", true),
        ("account-history.json", false),
    ];

    #[derive(err_derive::Error, Debug)]
    #[error(no_from)]
    pub enum Error {
        #[error(display = "Unable to find local appdata directory")]
        FindAppData,

        #[error(display = "Could not acquire security descriptor of backup directory")]
        SecurityInformation(#[error(source)] io::Error),

        #[error(display = "Backup directory is not owned by SYSTEM or Built-in Administrators")]
        WrongOwner,

        #[error(display = "Failed to copy files during migration")]
        Io(#[error(source)] io::Error),
    }

    /// Attempts to restore the Mullvad settings from `C:\windows.old` after an update of Windows.
    /// Upon success, it returns `Ok(true)` if the migration succeeded, and `Ok(false)` if no
    /// migration was needed.
    pub async fn migrate_after_windows_update(
        destination_settings_dir: &Path,
    ) -> Result<bool, Error> {
        let system_appdata_dir = dirs::data_local_dir().ok_or(Error::FindAppData)?;
        if !destination_settings_dir.starts_with(system_appdata_dir) {
            return Ok(false);
        }

        let settings_path = destination_settings_dir.join(super::SETTINGS_FILE);
        if settings_path.exists() {
            return Ok(false);
        }

        let mut components = destination_settings_dir.components();
        let prefix = if let Some(prefix) = components.next() {
            prefix
        } else {
            return Ok(false);
        };
        let root = if let Some(root) = components.next() {
            root
        } else {
            return Ok(false);
        };

        let windows_old_dir = Path::new(&prefix).join(root).join(MIGRATION_DIRNAME);
        let source_settings_dir = Path::new(&windows_old_dir).join(&components);
        if !source_settings_dir.exists() {
            return Ok(false);
        }

        let security_info =
            SecurityInformation::from_file(windows_old_dir.as_path(), OWNER_SECURITY_INFORMATION)
                .map_err(Error::SecurityInformation)?;

        let owner_sid = security_info.owner().ok_or(Error::WrongOwner)?;

        if !is_well_known_sid(owner_sid, WinLocalSystemSid)
            && !is_well_known_sid(owner_sid, WinBuiltinAdministratorsSid)
        {
            return Err(Error::WrongOwner);
        }

        if !destination_settings_dir.exists() {
            fs::create_dir_all(destination_settings_dir)
                .await
                .map_err(Error::Io)?;
        }

        let mut result = Ok(true);

        for (file, required) in &MIGRATE_FILES {
            let from = source_settings_dir.join(file);
            let to = destination_settings_dir.join(file);

            log::debug!("Migrating {} to {}", from.display(), to.display());

            match fs::copy(&from, &to).await {
                Ok(_) => {
                    let _ = fs::remove_file(from).await;
                }
                Err(error) => {
                    log::error!(
                        "{}",
                        error.display_chain_with_msg(&format!(
                            "Failed to copy {} to {}",
                            from.display(),
                            to.display()
                        ))
                    );
                    if *required {
                        result = Err(Error::Io(error));
                    }
                }
            }
        }

        if let Err(error) = fs::remove_dir(source_settings_dir).await {
            log::trace!(
                "{}",
                error.display_chain_with_msg("Failed to delete backup directory")
            );
        }

        result
    }

    struct SecurityInformation {
        security_descriptor: *mut SECURITY_DESCRIPTOR,
        owner: PSID,
    }

    impl SecurityInformation {
        pub fn from_file<T: AsRef<OsStr>>(
            path: T,
            security_information: SECURITY_INFORMATION,
        ) -> Result<Self, io::Error> {
            Self::from_object(path, SE_FILE_OBJECT, security_information)
        }

        pub fn from_object<T: AsRef<OsStr>>(
            object_name: T,
            object_type: SE_OBJECT_TYPE,
            security_information: SECURITY_INFORMATION,
        ) -> Result<Self, io::Error> {
            let mut u16_path: Vec<u16> = object_name.as_ref().encode_wide().collect();
            u16_path.push(0u16);

            let mut security_descriptor = ptr::null_mut();
            let mut owner = ptr::null_mut();

            let status = unsafe {
                GetNamedSecurityInfoW(
                    u16_path.as_ptr(),
                    object_type,
                    security_information,
                    &mut owner,
                    ptr::null_mut(),
                    ptr::null_mut(),
                    ptr::null_mut(),
                    &mut security_descriptor,
                )
            };

            if status != ERROR_SUCCESS {
                return Err(std::io::Error::from_raw_os_error(status as i32));
            }

            Ok(SecurityInformation {
                security_descriptor: security_descriptor as *mut _,
                owner,
            })
        }

        pub fn owner(&self) -> Option<&SID> {
            unsafe { (self.owner as *const SID).as_ref() }
        }
    }

    impl Drop for SecurityInformation {
        fn drop(&mut self) {
            unsafe { LocalFree(self.security_descriptor as HANDLE) };
        }
    }

    fn is_well_known_sid(sid: &SID, well_known_sid_type: WELL_KNOWN_SID_TYPE) -> bool {
        unsafe { IsWellKnownSid(sid as *const SID as *mut _, well_known_sid_type) == 1 }
    }
}
