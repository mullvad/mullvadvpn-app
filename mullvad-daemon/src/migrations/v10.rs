use super::{Error, Result};
use mullvad_types::settings::SettingsVersion;
use talpid_types::net::TunnelType;

/// Automatic tunnel protocol has been removed. If the tunnel protocol is set to `any`, it will be
/// migrated to `wireguard`, unless the location is an openvpn relay, in which case it will be
/// migrated to `openvpn`.
pub fn migrate(settings: &mut serde_json::Value) -> Result<()> {
    if !(version(settings) == Some(SettingsVersion::V10)) {
        return Ok(());
    }

    log::info!("Migrating settings format to v11");

    migrate_tunnel_type(settings)?;

    settings["settings_version"] = serde_json::json!(SettingsVersion::V11);

    Ok(())
}

fn version(settings: &serde_json::Value) -> Option<SettingsVersion> {
    settings
        .get("settings_version")
        .and_then(|version| serde_json::from_value(version.clone()).ok())
}

fn relay_settings(settings: &mut serde_json::Value) -> Option<&mut serde_json::Value> {
    settings.get_mut("relay_settings")?.get_mut("normal")
}

fn migrate_tunnel_type(settings: &mut serde_json::Value) -> Result<()> {
    if let Some(ref mut normal) = relay_settings(settings) {
        migrate_tunnel_type_inner(normal)?;
    }
    Ok(())
}

fn migrate_tunnel_type_inner(normal: &mut serde_json::Value) -> Result<()> {
    match normal.get_mut("tunnel_protocol") {
        // Migrate tunnel protocol "any"
        Some(serde_json::Value::String(s)) if s == "any" => {
            // If openvpn is selected, migrate to openvpn tunnel type
            // Otherwise, select wireguard
            let hostname = normal
                .get_mut("location")
                .and_then(|location| location.get_mut("only"))
                .and_then(|only| only.get_mut("location"))
                .and_then(|only| only.get_mut("hostname").cloned());

            let protocol = if let Some(serde_json::Value::String(s)) = hostname {
                if s.split('-').any(|token| token == "ovpn") {
                    TunnelType::OpenVpn
                } else {
                    TunnelType::Wireguard
                }
            } else {
                TunnelType::Wireguard
            };

            normal["tunnel_protocol"] = serde_json::json!(protocol);
        }
        // Migrate '"only": { "tunnel_protocol": $tunnel_protocol }'
        // to '"tunnel_protocol": $tunnel_protocol'
        Some(serde_json::Value::Object(ref mut constraint)) => {
            if let Some(tunnel_type) = constraint.get("only") {
                let tunnel_type: TunnelType = serde_json::from_value(tunnel_type.clone())
                    .map_err(|_| Error::InvalidSettingsContent)?;
                normal["tunnel_protocol"] = serde_json::json!(tunnel_type);
            } else {
                return Err(Error::InvalidSettingsContent);
            }
        }
        Some(_) => {
            return Err(Error::InvalidSettingsContent);
        }
        // Unexpected result. Do nothing.
        None => (),
    };
    Ok(())
}
#[cfg(test)]
mod test {
    use serde_json::json;

    use crate::migrations::v10::migrate_tunnel_type_inner;

    /// Tunnel protocol "any" is migrated to wireguard
    #[test]
    fn test_v10_to_v11_migration() {
        let mut old_settings = json!({
            "tunnel_protocol": "any",
            "location": {
                "only": {
                    "location": {
                        "country": "se"
                    }
                }
            }
        });
        migrate_tunnel_type_inner(&mut old_settings).unwrap();
        let new_settings: serde_json::Value = json!({
            "tunnel_protocol": "wireguard",
            "location": {
                "only": {
                    "location": {
                        "country": "se"
                    }
                }
            }
        });
        assert_eq!(&old_settings, &new_settings);
    }

    /// Tunnel protocol "any" is migrated to openvpn, since the location is an openvpn relay
    #[test]
    fn test_v10_to_v11_migration_openvpn_location() {
        let mut old_settings = json!({
            "tunnel_protocol": "any",
            "location": {
                "only": {
                    "location": {
                        "hostname": "at-vie-ovpn-001"
                    }
                }
            }
        });
        migrate_tunnel_type_inner(&mut old_settings).unwrap();
        let new_settings: serde_json::Value = json!({
            "tunnel_protocol": "openvpn",
            "location": {
                "only": {
                    "location": {
                        "hostname": "at-vie-ovpn-001"
                    }
                }
            }
        });
        assert_eq!(&old_settings, &new_settings);
    }

    /// Tunnel protocol '"only" = { "tunnel_protocol" = '$tunnel_protocol' } is migrated to
    /// '"tunnel_protocol" = '$tunnel_protocol'
    #[test]
    fn test_v10_to_v11_migration_only_protocol() {
        let mut old_settings = json!({
            "tunnel_protocol": {
                "only": "wireguard"
            },
            "location": {
                "only": {
                    "location": {
                        "country": "se"
                    }
                }
            }
        });
        migrate_tunnel_type_inner(&mut old_settings).unwrap();
        let new_settings: serde_json::Value = json!({
            "tunnel_protocol": "wireguard",
            "location": {
                "only": {
                    "location": {
                        "country": "se"
                    }
                }
            }
        });
        assert_eq!(&old_settings, &new_settings);
    }
}
