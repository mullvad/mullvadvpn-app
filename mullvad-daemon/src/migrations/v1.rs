use super::Result;
use mullvad_types::{constraints::Constraint, settings::SettingsVersion};
use serde::{Deserialize, Serialize};

// ======================================================
// Section for vendoring types and values that
// this settings version depend on. See `mod.rs`.

/// The tunnel protocol used by a [`TunnelEndpoint`][talpid_types::net::TunnelEndpoint].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename = "tunnel_type")]
pub enum TunnelType {
    #[serde(rename = "openvpn")]
    OpenVpn,
    #[serde(rename = "wireguard")]
    Wireguard,
}

// ======================================================

pub fn migrate(settings: &mut serde_json::Value) -> Result<()> {
    let version_matches = |settings: &serde_json::Value| settings.get("settings_version").is_none();
    if !version_matches(settings) {
        return Ok(());
    }

    log::info!("Migrating settings format to V2");

    let openvpn_constraints = || -> Option<serde_json::Value> {
        settings
            .get("relay_settings")?
            .get("normal")?
            .get("tunnel")?
            .get("only")?
            .get("openvpn")
            .cloned()
    }();
    let wireguard_constraints = || -> Option<serde_json::Value> {
        settings
            .get("relay_settings")?
            .get("normal")?
            .get("tunnel")?
            .get("only")?
            .get("wireguard")
            .cloned()
    }();

    if let Some(relay_settings) = settings.get_mut("relay_settings")
        && let Some(normal_settings) = relay_settings.get_mut("normal")
    {
        if let Some(openvpn_constraints) = openvpn_constraints {
            normal_settings["openvpn_constraints"] = openvpn_constraints;
            normal_settings["tunnel_protocol"] = serde_json::json!(Constraint::<TunnelType>::Any);
        } else if let Some(wireguard_constraints) = wireguard_constraints {
            normal_settings["wireguard_constraints"] = wireguard_constraints;
            normal_settings["tunnel_protocol"] =
                serde_json::json!(Constraint::Only(TunnelType::Wireguard));
        } else {
            normal_settings["tunnel_protocol"] = serde_json::json!(Constraint::<TunnelType>::Any);
        }
        if let Some(object) = normal_settings.as_object_mut() {
            object.remove("tunnel");
        }
    }

    settings["show_beta_releases"] = serde_json::json!(false);
    settings["settings_version"] = serde_json::json!(SettingsVersion::V2);

    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::migrations::load_seed;

    /// Parse example v1 settings as a pretty printed JSON string.
    fn v1_settings() -> serde_json::Value {
        load_seed("v1.json")
    }

    fn v1_2019v3_settings() -> serde_json::Value {
        load_seed("v1_2019v3.json")
    }

    #[test]
    fn snapshot_v1_settings() {
        let v1 = serde_json::to_string_pretty(&v1_settings()).unwrap();
        insta::assert_snapshot!(v1);
    }

    #[test]
    fn snapshot_v1_2019v3_settings() {
        let v1_2019v3 = serde_json::to_string_pretty(&v1_2019v3_settings()).unwrap();
        insta::assert_snapshot!(v1_2019v3);
    }

    #[test]
    fn v1_to_v2_migration() {
        let mut settings = load_seed("v1.json");
        migrate(&mut settings).unwrap();
        insta::assert_snapshot!(serde_json::to_string_pretty(&settings).unwrap());
    }

    #[test]
    fn v1_2019v3_to_v2_migration() {
        let mut settings = load_seed("v1_2019v3.json");
        migrate(&mut settings).unwrap();
        insta::assert_snapshot!(serde_json::to_string_pretty(&settings).unwrap());
    }
}
