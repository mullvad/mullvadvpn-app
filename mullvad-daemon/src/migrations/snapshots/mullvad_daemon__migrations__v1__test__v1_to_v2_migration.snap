---
source: mullvad-daemon/src/migrations/v1.rs
expression: "serde_json::to_string_pretty(&settings).unwrap()"
---
{
  "account_token": "1234",
  "allow_lan": true,
  "auto_connect": false,
  "block_when_disconnected": false,
  "relay_settings": {
    "normal": {
      "location": {
        "only": {
          "country": "se"
        }
      },
      "openvpn_constraints": {
        "port": {
          "only": 53
        },
        "protocol": {
          "only": "udp"
        }
      },
      "tunnel_protocol": "any"
    }
  },
  "settings_version": 2,
  "show_beta_releases": false,
  "tunnel_options": {
    "generic": {
      "enable_ipv6": false
    },
    "openvpn": {
      "mssfix": null
    },
    "wireguard": {
      "mtu": null
    }
  }
}
